File: leetcode/apple/two_sum.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> index;
        int half_index = -1;
        int val;
        for (int i=0; i<(int)nums.size(); i++) {
            val = nums[i];
            if (val*2 == target) {
                if (half_index != -1) {
                    return vector<int> {half_index, i};
                } else {
                    half_index = i;
                }
            }

            index[val] = i;
        }

        int comp;
        for (int i=0; i<(int)nums.size(); i++) {
            val = nums[i];
            comp = target - val;

            if ((index.find(comp) != index.end()) && (index[comp] != index[val])) {
                return vector<int> {i, index[comp]};
            }
        }

        return vector<int> {};
    }
};


int main() {
    vector<int> nums = {3, 2, 4};
    int target = 6;

    Solution S;
    vector<int> ans = S.twoSum(nums, target);
    cout << ans[0] << ' ' << ans[1] << endl;

    return 0;
}

File: leetcode/neetcode_roadmap/two_sum.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        nums = [(n, i) for i, n in enumerate(nums)]
        nums = sorted(nums, key=lambda x: x[0])

        i = 0
        j = len(nums) - 1
        while i < j:
            summed = nums[i][0] + nums[j][0]
            if summed == target:
                return [nums[i][1], nums[j][1]]

            if summed < target:
                i += 1
            else:
                j -= 1

S = Solution()

nums = [2,7,11,15]
target = 9
print(S.twoSum(nums, target))
